V2 : support sparse arrays from main()

z3dev opened the pull request:
As found by @gilboonet, designs that return sparse arrays, arrays filled with non-geometry, or arrays filled with empty geometry fail big time. These changes correct several packages to allow any combinaton of the above to be returned.

In addition to the fixes, the code that serializes solids to/from web worker content has been split out. So, only the core package understands / maintains this functionality.

Fixes #665 and a whole lot more.

The change to .gitignore is for VIM temporary files on Windows.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

Extensive tesing with the WEB UI was performed to verify that all functionality works, as well as the fixes above.

z3dev commented:
Test design what returns all kinds of crap.

```
const { path2, geom2, geom3 } = require('@jscad/modeling').geometries
const { square } = require('@jscad/modeling').primitives

const main = () => {
  let a = path2.create()
  let b = geom2.create()
  let c = geom3.create()
  let d = 'hello'
  let e = square()

  return [undefined, a, null, b, null, c, null, d, null, e, undefined]
}

module.exports = { main }
```

z3dev commented:
@SimonClark please review. This should be ready to merge.

SimonClark reviewed (approved):
:+1

SimonClark commented:
Sorry for the confusion.  I had reviewed, but not submitted the review.  Github's PR review process has changed a bit since I last used it a lot.

I won't merge yet, just in case you want to address the dead code.  Feel free to merge, or I'll do it later today, if you don't get to it.

z3dev commented on packages/utils/regl-renderer/src/geometry-utils-V2/entitiesFromSolids.js:
```diff
@@ -17,32 +17,37 @@ const entitiesFromSolids = (params, solids) => {
-    // geometry = flatten(geometries)// FXIME : ACTUALLY deal with arrays since a single design can
-    // generate multiple geometries if positions count is >65535
-    geometry = flatten(geometry)[0]
+    // FIXME handle multiple geometries, i.e. when positions count is > 65535
```
There’s a limit on the number of vertices handled by a single command in WEBGL. If a HUGE geometry is passed then something bad is going to happen, or something will be displayed completely.

I’m not interested in fixing this now. I’ll add a new issue to track this, and fix later.

z3dev commented on packages/core/code-evaluation/deserializeSolids.js:
```diff
@@ -0,0 +1,25 @@
+ */
+const deserializeSolids = (solids) => solids.map((solid) => {
+  if (!solid) return solid // handle passing of null / undefined
+  // if (solid[0] === 0) { // Geom2
```
I understand the motivation to serializer to buffers but the whole implementation is wrong. The buffers need to be passed specifically as transferable objects, which is not the case.

Therefore, the code was changed to serialize using JSON strings. And no information is lost. All is well.

Should the unused  code be removed? Yeah. I guess so.
Is there some way to document this issue for future reference?

z3dev commented:
OK. I removed the unused code, and added a comment about the use of compactBinary formats.

This should be ready for merging now. Thanks again @SimonClark

z3dev commented:
@SimonClark back to you for review. This should be ready to merge.

SimonClark reviewed (approved):
👍
